src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1381 @@
+                {
+                    analyzeBinary(currentCases, nestedBinary, wrapIntoParentAndPattern, context);
+                }
+            }
+
+            static void analyzeBinary(ArrayBuilder<BoundPattern> currentCases, BoundBinaryPattern binaryPattern, Func<BoundPattern, BoundPattern>? wrapIntoParentAndPattern, ReachabilityAnalysisContext context)
+            {
+                if (binaryPattern.Disjunction)
+                {
+                    int savedStackCount = currentCases.Count;
+
+                    var patterns = ArrayBuilder<BoundPattern>.GetInstance();
+                    addPatternsFromOrTree(binaryPattern, patterns);
+
+                    // In `A1 or ... or Ai or B or ...`, we analyze `B` with `case A1:`, ..., `case Ai:` (with each wrapped as indicated by caller) added to current cases.
+                    // That way, if `B` can be seen as multiple cases: `case B1`, ..., `case Bn`
+                    // we'll be able to check reachability on: `case A1:`, ... `case Ai:`, `case B1:`, ... `case Bn:`
+                    for (int i = 0; i < patterns.Count; i++)
+                    {
+                        analyzePattern(currentCases, patterns[i], wrapIntoParentAndPattern, context);
+                        var wrappedPattern = wrapIntoParentAndPattern?.Invoke(patterns[i]) ?? patterns[i];
+                        currentCases.Add(wrappedPattern);
+                    }
+
+                    // For `A1 or ... or An`, we check reachability on: `case A1`, ..., `case An` (with each wrapped as indicated by caller)
+                    checkReachability(currentCases, context);
+
+                    currentCases.Count = savedStackCount;
+                    patterns.Free();
+                }
+                else
+                {
+                    var stack = ArrayBuilder<BoundBinaryPattern>.GetInstance();
+                    BoundBinaryPattern? current = binaryPattern;
+                    do
+                    {
+                        stack.Push(current);
+                        current = current.Left as BoundBinaryPattern;
+                    }
+                    while (current != null && !current.Disjunction);
+
+                    current = stack.Pop();
+                    // In `A and B and ...`, we analyze `A` without the `and B and ...`
+                    analyzePattern(currentCases, current.Left, wrapIntoParentAndPattern, context);
+
+                    do
+                    {
+                        // In `A and B`, we analyze `B` but any cases `case B1:`, ..., `case Bn:` found there
+                        // will be wrapped as `A and <expansion>`.
+                        // So we'll check reachability on `case A and B1:`, ..., `case A and Bn:`
+
+                        // Given `newPattern`, produce `A and newPattern`
+                        Func<BoundPattern, BoundPattern> newWrapIntoParentAndPattern = (BoundPattern newPattern) =>
```

Comment:
>newWrapIntoParentAndPattern [](http://example.com/codeflow?start=57&length=27)

It looks like we can decrease amount of delegates that we create by moving this declaration out of the loop. It doesn't capture anything declared in the loop.